lib/srv/app/server.go

```diff
@@ -0,0 +1,501 @@
+	clusterConfig, err := s.c.AccessPoint.GetClusterConfig()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	s.keepAlive = clusterConfig.GetKeepAliveInterval()
+
+	// Figure out the port the proxy is running on.
+	s.proxyPort = s.getProxyPort()
+
+	return s, nil
+}
+
+// GetServerInfo returns a services.Server representing the application. Used
+// in heartbeat code.
+func (s *Server) GetServerInfo() (services.Server, error) {
+	// Update dynamic labels on all apps.
+	apps := s.c.Server.GetApps()
+	for _, a := range apps {
+		dl, ok := s.dynamicLabels[a.Name]
+		if !ok {
+			continue
+		}
+		a.DynamicLabels = services.LabelsToV2(dl.Get())
+	}
+	s.c.Server.SetApps(apps)
+
+	// Update the TTL.
+	s.c.Server.SetTTL(s.c.Clock, defaults.ServerAnnounceTTL)
+
+	// Update rotation state.
+	rotation, err := s.c.GetRotation(teleport.RoleApp)
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			s.log.Warningf("Failed to get rotation state: %v.", err)
+		}
+	} else {
+		s.c.Server.SetRotation(*rotation)
+	}
+
+	return s.c.Server, nil
+}
+
+// Start starts heart beating the presence of service.Apps that this
+// server is proxying along with any dynamic labels.
+func (s *Server) Start() {
+	for _, dynamicLabel := range s.dynamicLabels {
+		go dynamicLabel.Start()
+	}
+	go s.heartbeat.Run()
+}
+
+// Close will shut the server down and unblock any resources.
+func (s *Server) Close() error {
+	var errs []error
+
+	// Stop HTTP server.
+	if err := s.httpServer.Close(); err != nil {
+		errs = append(errs, err)
+	}
+
+	// Stop heartbeat to auth.
+	if err := s.heartbeat.Close(); err != nil {
+		errs = append(errs, err)
+	}
+
+	// Stop all dynamic labels from being updated.
+	for _, dynamicLabel := range s.dynamicLabels {
+		dynamicLabel.Close()
+	}
+
+	// Signal to any blocking go routine that it should exit.
+	s.closeFunc()
+
+	return trace.NewAggregate(errs...)
+}
+
+// Wait will block while the server is running.
+func (s *Server) Wait() error {
+	<-s.closeContext.Done()
+	return s.closeContext.Err()
+}
+
+// ForceHeartbeat is used in tests to force updating of services.Server.
+func (s *Server) ForceHeartbeat() error {
+	err := s.heartbeat.ForceSend(time.Second)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// HandleConnection takes a connection and wraps it in a listener so it can
+// be passed to http.Serve to process as a HTTP request.
+func (s *Server) HandleConnection(conn net.Conn) {
+	// Wrap the listener in a TLS authorizing listener.
+	listener := newListener(s.closeContext, conn)
+	tlsListener := tls.NewListener(listener, s.tlsConfig)
+
+	if err := s.httpServer.Serve(tlsListener); err != nil {
+		s.log.Warnf("Failed to handle connection: %v.", err)
+	}
+}
+
+// ServeHTTP will forward the *http.Request to the target application.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if err := s.serveHTTP(w, r); err != nil {
+		s.log.Debugf("Failed to serve request: %v.", err)
```

Comment:
Wouldn't it be useful to see this error in the logs without having to enable debug mode?